src/irmin-pack/mem/irmin_pack_mem.ml

```diff
@@ -0,0 +1,158 @@
+        include Irmin.Contents.Store (CA) (H) (C)
+      end
+
+      module Node = struct
+        module Node = Node (H) (P) (M)
+
+        module CA = struct
+          module Inter = Irmin_pack.Inode.Make_internal (Config) (H) (Node)
+          module CA = Pack.Make (Inter.Raw)
+          include Irmin_pack.Inode.Make (H) (Node) (Inter) (CA)
+
+          let v = CA.v
+        end
+
+        include Irmin.Private.Node.Store (Contents) (CA) (H) (CA.Val) (M) (P)
+      end
+
+      module Commit = struct
+        module Commit = Commit.Make (H)
+        module Pack_value = Irmin_pack.Pack_value.Of_commit (H) (Commit)
+        module CA = CA_mem (H) (Pack_value)
+        include Irmin.Private.Commit.Store (Info) (Node) (CA) (H) (Commit)
+      end
+
+      module Branch = struct
+        module Key = B
+        module Val = H
+        module AW = Atomic_write (Key) (Val)
+        include Irmin_pack.Atomic_write.Closeable (AW)
+
+        let v () = AW.v () >|= make_closeable
+      end
+
+      module Slice = Irmin.Private.Slice.Make (Contents) (Node) (Commit)
+      module Remote = Irmin.Private.Remote.None (H) (B)
+
+      module Repo = struct
+        type t = {
+          config : Irmin.Private.Conf.t;
+          contents : read Contents.CA.t;
+          node : read Node.CA.t;
+          commit : read Commit.CA.t;
+          branch : Branch.t;
+        }
+
+        let contents_t t : 'a Contents.t = t.contents
+        let node_t t : 'a Node.t = (contents_t t, t.node)
+        let commit_t t : 'a Commit.t = (node_t t, t.commit)
+        let branch_t t = t.branch
+
+        let batch t f =
+          Commit.CA.batch t.commit (fun commit ->
+              Node.CA.batch t.node (fun node ->
+                  Contents.CA.batch t.contents (fun contents ->
+                      let contents : 'a Contents.t = contents in
+                      let node : 'a Node.t = (contents, node) in
+                      let commit : 'a Commit.t = (node, commit) in
+                      f contents node commit)))
+
+        let v config =
+          let root = Irmin_pack.Conf.root config in
+          let* contents = Contents.CA.v root in
+          let* node = Node.CA.v root in
+          let* commit = Commit.CA.v root in
+          let+ branch = Branch.v () in
+          { contents; node; commit; branch; config }
+
+        let close t =
+          Contents.CA.close (contents_t t) >>= fun () ->
+          Node.CA.close (snd (node_t t)) >>= fun () ->
+          Commit.CA.close (snd (commit_t t)) >>= fun () -> Branch.close t.branch
+
+        (* An in-memory store is always in sync. *)
+        let sync _ = ()
+        let flush _ = ()
+
+        (* Stores share instances so one clear is enough. *)
+        let clear t = Contents.CA.clear (contents_t t)
+      end
+    end
+
+    include Irmin.Of_private (X)
+
+    let integrity_check_inodes ?heads:_ _ =
+      Lwt.return (Ok (`Msg "No inodes within an in-memory store"))
```

Comment:
This should raise "not implemented", or re-implemented: https://github.com/mirage/irmin/commit/245c9174d358cce2a6d2f2b902004b750d2cb766

Reply:
I'll go with not implemented for now, as this is exposed only for the CLI. We can factor out your re-implementation later if we find a use-case.